### PR TITLE
Babel plugin

### PR DESCRIPTION
**Pull request by cangarugula**
Opened up for your thoughts on implementation and if this the direction you were thinking @threepointone 

This is a babel plugin that scans files that import iff and call it with a flag name. It outputs a JSON file that holds the count of each flag name used in the code base. `yarn iff-scan` will execute this scan.

I added an example of setting __iff_VALUES__ and calling iff within a component. The iffValues.json file should contain "featureflagging.in.cib.enable": 1

Several items to do left:
- Determine whether to include it within this project or as its own repo
- Silence the compiled code logs when running `iff-scan`

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -1,4 +1,7 @@
 {
   "extends": "modular-scripts/tsconfig.json",
-  "include": ["modular", "packages/*/src"]
-}
+  "include": ["modular", "packages/*/src"],
+  "compilerOptions": {
+    "isolatedModules": false
```
Comment (threepointone) [inline tsconfig.json:5]: isn't isolateModules already true by default? https://github.com/jpmorganchase/modular/blob/406fa0a740cc511d8b504326150ee182e035c571/packages/modular-scripts/tsconfig.json#L16

File: packages/iff/src/server.ts [hunk 2]
```diff
@@ -2,7 +2,11 @@ import { AsyncLocalStorage } from 'async_hooks';
 
 export const flagLocalStorage = new AsyncLocalStorage();
 
-export default function iff(flagName: string, defaultValue: any) {
+export default function iff(flagName: string) {
+  if (typeof flagName !== 'string') {
+    throw TypeError('flagName must be a string');
+  }
```
Comment (threepointone) [inline packages/iff/src/server.ts:8]: you shouldn't need this check, typescript should catch it automatically.

File: packages/iff/src/server.ts [hunk 3]
```diff
@@ -11,7 +15,7 @@ export default function iff(flagName: string, defaultValue: any) {
   }
 
   if (flags[flagName] === undefined) {
-    return defaultValue;
+    return null;
```
Comment (threepointone) [inline packages/iff/src/server.ts:18]: let's just return undefined instead, null could be a valid value set on the service. so remove this if block completely.

File: packages/iff/src/index.ts [hunk 4]
```diff
@@ -1,6 +1,2 @@
-export default function iff<T>(flagName: string, defaultValue?: T) {
-  // test once if values have been preloaded
-  // freeze
-  // in dev, make sure it's shallow?
-  console.log('soon!');
-}
+export { default } from './browser';
+// since we are only offering iff as a browser package, we should set the browser package as the default export for iff
```
Comment (threepointone) [inline packages/iff/src/index.ts:2]: how about we bring the entire implementation from browser into this file, and delete browser.js?

File: packages/iff/src/browser.ts [hunk 5]
```diff
@@ -1,12 +1,19 @@
...
+  if (typeof flagName !== 'string') {
+    throw new TypeError('flagName must be a string');
+  }
   if (flags === undefined) {
     throw new Error('feature flags were not populated');
   }
 
   if (flags[flagName] === undefined) {
-    return defaultValue;
+    return null;
```
Comment (threepointone) [inline packages/iff/src/browser.ts:16]: delete this entire if block

File: packages/iff/src/browser.ts [hunk 6]
```diff
@@ -1,12 +1,19 @@
-export default function iff(flagName: string, defaultValue?: any) {
-  // @ts-expect-error
-  const flags = window.__iff_VALUES___;
+declare global {
+  interface Window {
+    __iff_VALUES__: any;
```
Comment (threepointone) [inline packages/iff/src/browser.ts:3]: we shouldn't add this type on to window, or else other folks will start accessing it. instead, add a `@tes-expect-error` below where we access __iff_VALUES__

File: package.json [hunk 7]
```diff
@@ -11,11 +11,12 @@
     "type": "root"
   },
   "scripts": {
-    "start": "modular start app",
+    "start": "NODE_PATH=packages/ modular start app",
```
Comment (threepointone) [inline package.json:14]: why did you need to add NODE_PATH here? is `iff` not importable from inside `app`?

File: package.json [hunk 8]
```diff
@@ -11,11 +11,12 @@
...
     "build": "modular build app",
-    "test": "modular test",
+    "test": "modular test ",
```
Comment (threepointone) [inline package.json:16]:
```suggestion
    "test": "modular test",
```

File: package.json [hunk 9]
```diff
@@ -35,6 +42,7 @@
     "@types/react": "^17.0.0",
     "@types/react-dom": "^17.0.0",
     "eslint-config-modular-app": "^0.3.1",
+    "fs": "^0.0.1-security",
```
Comment (threepointone) [inline package.json:45]: could you explain this?

File: .gitignore [hunk 10]
```diff
@@ -23,4 +23,7 @@ yarn-debug.log*
 yarn-error.log*
 
 .eslintcache
-dist
+/dist
+
+iffValues.json
+__snapshots__
```
Comment (threepointone) [inline .gitignore:29]: don't ignore snapshots, commit them to the repo

File: packages/app/src/__tests__/App.test.tsx [hunk 11]
```diff
@@ -3,6 +3,7 @@ import { render } from '@testing-library/react';
 import App from '../App';
 
 test('renders learn react link', () => {
+  window.__iff_VALUES__ = {};
```
Comment (threepointone) [inline packages/app/src/__tests__/App.test.tsx:6]: unset this value after the test ends 
(it should probably in a beforeEach/afterEach block, I think)

File: packages/iff/src/__tests__/babel-plugin.test.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+const resetOutputSourceFile = () => {
+  if (fs.existsSync(outputSourceFilePath)) {
+    fs.rmSync(outputSourceFilePath);
+  }
+};
+
+describe('iff scan babel plugin', () => {
+  beforeAll(() => {
+    if (fs.existsSync(path.join(__dirname, '__snapshots__'))) {
```
Comment (threepointone) [inline packages/iff/src/__tests__/babel-plugin.test.ts:28]: don't do this, we should use the snapshot to verify that values don't change.

File: packages/iff/src/__tests__/babel-plugin.test.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
+import { transform, BabelFileResult } from '@babel/core';
```
Comment (threepointone) [inline packages/iff/src/__tests__/babel-plugin.test.ts:1]: so, I don't think we should be transforming the code and generating transformed code. All we need to do is parse the code and spot `iff()` calls. Let's talk about this whole file tomorrow and try a second pass at it.

File: packages/iff/src/__tests__/browser.test.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+const originalWindow = window;
+
+describe('iff browser', () => {
+  const testFlag = 'feature-flagging-enable';
+  describe('when flagName is not a string type', () => {
+    it('should throw a type error', () => {
+      // @ts-expect-error
+      expect(() => iff({ [testFlag]: true })).toThrow(
+        new TypeError('flagName must be a string'),
+      );
```
Comment (threepointone) [inline packages/iff/src/__tests__/browser.test.ts:13]: we shouldn'r have to do it this way. we should instead make an eslint plugin/rule that detects and errors when a non-string literal is passed to iff(). Let's talk about this too tomorrow.

**Review by threepointone — COMMENTED** [review]
I see what you're trying to do here, and it's a dope attempt. I think we can simplify the approach drastically. Let's chat tomorrow and work on it some more. Thank you very much for starting the PR!

File: packages/iff/src/__tests__/babel-plugin.test.ts (hunk 12 above)
Comment (cangarugula) [reply packages/iff/src/__tests__/babel-plugin.test.ts:28]: If we change any of the fixtures files down the line, we'd have to manually delete the snapshot folder before running the tests to reset the snapshots so I figured that this would teardown previous tests' snapshots to start anew.

File: packages/iff/src/__tests__/babel-plugin.test.ts (hunk 13 above)
Comment (cangarugula) [reply packages/iff/src/__tests__/babel-plugin.test.ts:1]: This transform function is only used for testing the plugin against code passed in, it's not actually transforming anything. It was the recommended way to test babel plugins. I have a specific unit test that checks that it is not transforming any code on line 38

File: packages/iff/src/server.ts (hunk 2 above)
Comment (cangarugula) [reply packages/iff/src/server.ts:8]: Even if the project that uses iff is a JS project, not TS? Probably a gap in TS knowledge here.

File: package.json (hunk 7 above)
Comment (cangarugula) [reply package.json:14]: I wanted to set up app so that it would import 'iff' as if it were a dependency so that the babel-plugin would recognize as it as a module import and not as a path so I set NODE_PATH to point to the folder to mimic this. I assumed that the app inside iff is there to test out its use but let me know if I misunderstood.
